File: packages/@paperclip-lang/core/src/parser/dsl/tokenizer.ts
import { negate } from "lodash";
import { UnknownTokenError } from "../base/errors";
import { BaseToken } from "../base/state";
import {
  BaseTokenizer,
  scanNewLine,
  scanNumberValue,
  scanString,
  scanWhitespace,
  token,
} from "../base/tokenizer";
import {
  isDigit,
  isLetter,
  isLetterOrDigit,
  isNewLine,
  isSpace,
  isWhitespace,
} from "../base/utils";

export enum DSLTokenKind {
  Keyword = 1 << 1,
  Dot = 1 << 2,
  Number = 1 << 3,
  CurlyOpen = 1 << 4,
  CurlyClose = 1 << 5,
  ParenOpen = 1 << 6,
  ParenClose = 1 << 7,
  Colon = 1 << 8,
  SquareOpen = 1 << 9,
  SquareClose = 1 << 10,
  Boolean = 1 << 11,

  // tokenizer handles this since it's much more efficient
  MultiLineComment = 1 << 12,
  SingleLineComment = 1 << 13,
  Comma = 1 << 14,
  At = 1 << 15,
  String = 1 << 16,
  Space = 1 << 17,
  NewLine = 1 << 18,
  Minus = 1 << 19,
  Percent = 1 << 20,
  Pound = 1 << 21,
  Char = 1 << 22,
}

export const DSL_SUPERFLUOUS_TOKENS =
  DSLTokenKind.Space |
  DSLTokenKind.NewLine |
  DSLTokenKind.SingleLineComment |
  DSLTokenKind.MultiLineComment;

export const DSL_WHITESPACE_TOKENS = DSLTokenKind.Space | DSLTokenKind.NewLine;

export type Token = BaseToken<DSLTokenKind>;

export class DSLTokenizer extends BaseTokenizer<DSLTokenKind> {
  _next(): Token {
    const chr = this._scanner.currChar();
    this._scanner.nextChar();

    if (chr === "(") {
      return token(DSLTokenKind.ParenOpen, chr);
    }

    if (chr === ")") {
      return token(DSLTokenKind.ParenClose, chr);
    }

    if (chr === "{") {
      return token(DSLTokenKind.CurlyOpen, chr);
    }

    if (chr === "}") {
      return token(DSLTokenKind.CurlyClose, chr);
    }

    if (chr === ",") {
      return token(DSLTokenKind.Comma, chr);
    }

    if (chr === ":") {
      return token(DSLTokenKind.Colon, chr);
    }

    if (chr === "[") {
      return token(DSLTokenKind.SquareOpen, chr);
    }

    if (chr === "]") {
      return token(DSLTokenKind.SquareClose, chr);
    }

    if (chr === ".") {
      return token(DSLTokenKind.Dot, chr);
    }

    if (chr === "@") {
      return token(DSLTokenKind.At, chr);
    }

    if (chr === "%") {
      return token(DSLTokenKind.Percent, chr);
    }

    if (chr === "#") {
      return token(DSLTokenKind.Pound, chr);
    }

    if (chr === "/") {
      const curr = this._scanner.currChar();
      if (curr === "*") {
        this._scanner.nextChar(); // eat *
        let buffer = "/*";
        while (!this._scanner.isEOF()) {
          const curr = this._scanner.currChar();
          if (curr === "*" && this._scanner.peekChar() === "/") {
            buffer += "*/";
            this._scanner.skip(2);
            break;
          }
          buffer += curr;
          this._scanner.nextChar();
        }
        return token(DSLTokenKind.MultiLineComment, buffer);
      } else if (curr === "/") {
        this._scanner.nextChar(); // eat *
        let buffer = "//" + this._scanner.scanUntil((c) => c === "\n");

        // maybe EOF
        if (this._scanner.currChar() === "\n") {
          this._scanner.nextChar(); // eat \n
          buffer += "\n";
        }
        return token(DSLTokenKind.SingleLineComment, buffer);
      }
    }

    if (chr === '"' || chr === "'") {
      return token(DSLTokenKind.String, scanString(this._scanner, chr));
    }

    if (chr === "-") {
      if (isDigit(this._scanner.currChar())) {
        return token(DSLTokenKind.Number, chr + scanNumberValue(this._scanner));
      }
      return token(DSLTokenKind.Minus, chr);
    }

    if (isNewLine(chr)) {
      return token(DSLTokenKind.NewLine, chr);
    }

    if (isSpace(chr)) {
      return token(DSLTokenKind.Space, chr);
    }

    if (isLetter(chr)) {
      const keyword =
        chr +
        this._scanner.scanUntil(
          negate((c) => isLetter(c) || isDigit(c) || c === "-" || c === "_")
        );

      if (keyword === "true" || keyword === "false") {
        return token(DSLTokenKind.Boolean, keyword);
      }

      return token(DSLTokenKind.Keyword, keyword);
    }

    if (isDigit(chr)) {
      return token(DSLTokenKind.Number, chr + scanNumberValue(this._scanner));
    }

    return token(DSLTokenKind.Char, chr);
  }
}

File: packages/@paperclip-lang/core/native/src/pc/tokenizer.rs
// Inspired by https://github.com/servo/rust-cssparser/blob/master/src/tokenizer.rs
use crate::base::ast::{Range, U16Position};
use crate::base::tokenizer::next_scanner_char;
use crate::core::errors as err;
use crate::core::string_scanner::{
    is_az, is_digit, is_newline, is_space, scan_string, Char, StringScanner,
};
use serde::Serialize;

#[derive(Debug, PartialEq, Serialize, Clone)]
pub enum Token<'src> {
    None,
    CurlyOpen,
    CurlyClose,
    ParenOpen,
    ParenClose,
    SquareOpen,
    SquareClose,
    Comma,
    Colon,
    Backslack,
    Dot,
    NewLine(&'src [u8]),
    MultiLineComment(&'src [u8]),
    DoccoStart,
    Space(&'src [u8]),
    Number(&'src [u8]),
    Word(&'src [u8]),
    KeywordExtends,
    String(&'src [u8]),
    Byte(u8),

    // Possibly glyphs
    Cluster(&'src [u8]),
}

pub fn next_token<'src>(source: &mut StringScanner<'src>) -> Result<Token<'src>, err::ParserError> {
    let s_pos = source.get_pos();
    let c = next_scanner_char(source)?;

    match c {
        Char::Byte(b) => {
            Ok(match b {
                b'{' => Token::CurlyOpen,
                b'}' => Token::CurlyClose,
                b'[' => Token::SquareOpen,
                b']' => Token::SquareClose,
                b'(' => Token::ParenOpen,
                b')' => Token::ParenClose,
                b',' => Token::Comma,
                b':' => Token::Colon,
                b'/' => {
                    if source.peek(0) == Some(b'*') {
                        // eat *
                        source.forward(1);

                        // Look for /**
                        if source.peek(0) == Some(b'*') {
                            source.forward(1);
                            Token::DoccoStart
                        } else {
                            next_comment(source, s_pos.u8_pos)?
                        }
                    } else {
                        Token::Backslack
                    }
                }
                b'\"' | b'\'' => Token::String(scan_string(source, b)),
                _ if is_space(b) => {
                    let e_pos = source.scan(is_space).u8_pos;
                    Token::Space(&source.source[s_pos.u8_pos..e_pos])
                }
                _ if is_newline(b) => {
                    let e_pos = source.scan(is_newline).u8_pos;
                    Token::NewLine(&source.source[s_pos.u8_pos..e_pos])
                }
                _ if is_az(b) => {
                    let e_pos = source.scan(|b| is_az(b) || is_digit(b)).u8_pos;
                    let word = &source.source[s_pos.u8_pos..e_pos];
                    if word == b"extends" {
                        Token::KeywordExtends
                    } else {
                        Token::Word(word)
                    }
                }
                b'.' => Token::Dot,
                b => Token::Byte(b),
            })
        }
        Char::Cluster(value) => Ok(Token::Cluster(value)),
        _ => Err(err::ParserError::new(
            "Unexpected token".to_string(),
            Range::new(s_pos.to_u16(), source.get_u16pos()),
            err::ErrorKind::UnexpectedToken,
        )),
    }
}

fn next_comment<'src>(
    source: &mut StringScanner<'src>,
    s_pos: usize,
) -> Result<Token<'src>, err::ParserError> {
    // Eat the entire comment
    while !source.is_eof() {
        let curr = source.source[source.pos];
        if curr == b'*' && source.peek(1) == Some(b'/') {
            let e_pos = source.pos + 2;
            source.forward(2);
            return Ok(Token::MultiLineComment(&source.source[s_pos..e_pos]));
        }
        source.forward(1);
    }

    Err(err::ParserError::new_eof())
}

pub fn is_superfluous(token: &Token) -> bool {
    matches!(token, Token::Space(_) | Token::MultiLineComment(_))
}

pub fn is_superfluous_or_newline(token: &Token) -> bool {
    is_superfluous(token) || matches!(token, Token::NewLine(_))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_tokenize_multiline_comment() {
        let mut scanner = StringScanner::new("hello/*gfdfdsfsd*//**abba");
        assert_eq!(next_token(&mut scanner), Ok(Token::Word(b"hello")));
        assert_eq!(
            next_token(&mut scanner),
            Ok(Token::MultiLineComment(b"/*gfdfdsfsd*/"))
        );
        assert_eq!(next_token(&mut scanner), Ok(Token::DoccoStart));
        assert_eq!(next_token(&mut scanner), Ok(Token::Word(b"abba")));
    }
}
